ddl/session_pool.go

```diff
@@ -53,17 +62,21 @@ func (sg *sessionPool) put(ctx sessionctx.Context) {
 		return
 	}
 
+	// no need to protect sg.resPool, even the sg.resPool is closed, the ctx still need to
+	// put into resPool, because when resPool is closing, it will wait all the ctx returns, then resPool finish closing.
 	sg.resPool.Put(ctx.(pools.Resource))
 }
 
 // close clean up the sessionPool.
 func (sg *sessionPool) close() {
 	sg.mu.Lock()
 	defer sg.mu.Unlock()
-	if sg.resPool == nil {
+	// prevent closing resPool twice.
+	if sg.mu.closed || sg.resPool == nil {
 		return
 	}
 
+	log.Info("[ddl] closing sessionPool")
 	sg.resPool.Close()
-	sg.resPool = nil
+	sg.mu.closed = true
```

Comment:
Since you change here, should we add check `closed` when `get` and `put`?
em... Actually， I can not understand why you add `closed` variable, It seem redundant.

Reply:
closed is used to indicate the sessionPool is closed and avoid to close it twice. If we use sg.resPool == nil to check, it will cause concurrent access to sg.resPool, and we must use mutex to protect it. Actually *pools.ResourcePool is thread-safe. And if the closed is true, the sg.resPool.get() will return the error, and sg.resPool.put is still need to put the resource back to ctxPool, it is by design.

Reply:
I got you.